sklearn/cluster/_hdbscan/_boruvka.pyx

```diff
@@ -0,0 +1,760 @@
+        The Boruvka algorithm benefits from a smaller leaf size than
+        standard kd-tree nearest neighbor searches. The tree passed in
+        is used for a kNN search for core distance. A second tree is
+        constructed with a smaller leaf size for Boruvka; this is that
+        leaf size.
+
+    alpha : float, optional (default=1.0)
+        The alpha distance scaling parameter as per Robust Single Linkage.
+
+    approx_min_span_tree : bool, optional (default=False)
+        Take shortcuts and only approximate the min spanning tree.
+        This is considerably faster but does not return a true
+        minimal spanning tree.
+
+    n_jobs : int, optional (default=4)
+        The number of parallel jobs used to compute core distances.
+
+    **kwargs :
+        Keyword args passed to the metric.
+    """
+
+    cdef:
+        object tree
+        DistanceMetric64 dist
+        readonly const float64_t[:, ::1] raw_data
+        float64_t[:, :, ::1] node_bounds
+        float64_t alpha
+        int8_t approx_min_span_tree
+        intp_t n_jobs, min_samples
+        intp_t num_points, num_nodes, num_features
+        bint has_KDTree
+
+        float64_t[::1] core_distance
+        float64_t[::1] bounds
+        intp_t[::1] components
+        intp_t[::1] component_of_point
+        intp_t[::1] component_of_node
+        intp_t[::1] candidate_neighbor
+        intp_t[::1] candidate_point
+        float64_t[::1] candidate_distance
+        float64_t[:, ::1] centroid_distances
+        intp_t[::1] idx_array
+        NodeData_t[::1] node_data
+        BoruvkaUnionFind component_union_find
+        MST_edge_t[::1] edges
+        intp_t num_edges
+
+    def __init__(
+        self,
+        tree,
+        min_samples=5,
+        metric='euclidean',
+        leaf_size=20,
+        alpha=1.0,
+        approx_min_span_tree=False,
+        n_jobs=None,
+        **kwargs
+    ):
+
+        self.tree =tree
+        self.has_KDTree = isinstance(tree, KDTree)
+        self.raw_data = self.tree.data
+        self.node_bounds = self.tree.node_bounds
+        self.alpha = alpha
+        self.approx_min_span_tree = approx_min_span_tree
+        self.n_jobs = effective_n_jobs(n_jobs)
+        self.min_samples = min_samples
+
+        self.num_points = self.tree.data.shape[0]
+        self.num_features = self.tree.data.shape[1]
+        self.num_nodes = self.tree.node_data.shape[0]
+
+        self.dist = DistanceMetric.get_metric(metric, **kwargs)
+
+        self.components = np.arange(self.num_points, dtype=np.intp)
+        self.bounds = np.empty(self.num_nodes, np.float64)
+        self.component_of_point = np.empty(self.num_points, dtype=np.intp)
+        self.component_of_node = np.empty(self.num_nodes, dtype=np.intp)
+        self.candidate_neighbor = np.empty(self.num_points, dtype=np.intp)
+        self.candidate_point = np.empty(self.num_points, dtype=np.intp)
+        self.candidate_distance = np.empty(self.num_points, dtype=np.float64)
```

Comment:
Instead of using `_initialize_components` to fill them with for loops, we could use `np.full` here instead.